File: code/widgets/org.eclipse.scout.widgets.client/src/main/java/org/eclipse/scout/widgets/client/ui/forms/hybrid/PersonDo.java
/*
 * Copyright (c) 2022 BSI Business Systems Integration AG.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Distribution License v1.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/org/documents/edl-v10.html
 *
 * Contributors:
 *     BSI Business Systems Integration AG - initial API and implementation
 */
package org.eclipse.scout.widgets.client.ui.forms.hybrid;

import java.util.Date;

import javax.annotation.Generated;

import org.eclipse.scout.rt.dataobject.DoEntity;
import org.eclipse.scout.rt.dataobject.DoValue;
import org.eclipse.scout.rt.dataobject.TypeName;

@TypeName("widgets.Person")
public class PersonDo extends DoEntity {

  public DoValue<String> name() {
    return doValue("name");
  }

  public DoValue<String> firstName() {
    return doValue("firstName");
  }

  public DoValue<Date> dateOfBirth() {
    return doValue("dateOfBirth");
  }

  public DoValue<Boolean> active() {
    return doValue("active");
  }

  /* **************************************************************************
   * GENERATED CONVENIENCE METHODS
   * *************************************************************************/

  @Generated("DoConvenienceMethodsGenerator")
  public PersonDo withName(String name) {
    name().set(name);
    return this;
  }

  @Generated("DoConvenienceMethodsGenerator")
  public String getName() {
    return name().get();
  }

  @Generated("DoConvenienceMethodsGenerator")
  public PersonDo withFirstName(String firstName) {
    firstName().set(firstName);
    return this;
  }

  @Generated("DoConvenienceMethodsGenerator")
  public String getFirstName() {
    return firstName().get();
  }

  @Generated("DoConvenienceMethodsGenerator")
  public PersonDo withDateOfBirth(Date dateOfBirth) {
    dateOfBirth().set(dateOfBirth);
    return this;
  }

  @Generated("DoConvenienceMethodsGenerator")
  public Date getDateOfBirth() {
    return dateOfBirth().get();
  }

  @Generated("DoConvenienceMethodsGenerator")
  public PersonDo withActive(Boolean active) {
    active().set(active);
    return this;
  }

  @Generated("DoConvenienceMethodsGenerator")
  public Boolean getActive() {
    return active().get();
  }

  @Generated("DoConvenienceMethodsGenerator")
  public boolean isActive() {
    return nvl(getActive());
  }
}

File: code/widgets/org.eclipse.scout.widgets.ui.html/src/main/js/forms/hybrid/PersonDo.ts
/*
 * Copyright (c) 2010, 2023 BSI Business Systems Integration AG
 *
 * This program and the accompanying materials are made
 * available under the terms of the Eclipse Public License 2.0
 * which is available at https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 */

import {dates, objects, strings} from '@eclipse-scout/core';

export class PersonDo {
  _type: 'widgets.Person';
  name: string;
  firstName: string;
  dateOfBirth: Date;
  active: boolean;

  constructor() {
    this._type = 'widgets.Person';
  }

  toJson(): object {
    const result = {};
    Object.keys(this).forEach(key => (result[key] = this._propertyToJson(this[key])));
    return result;
  }

  protected _propertyToJson(property: any): any {
    if (property instanceof Date) {
      return dates.toJsonDate(property);
    }
    return property;
  }

  static of(stringOrObject: string | object): PersonDo {
    if (!stringOrObject) {
      return null;
    }
    if (objects.isString(stringOrObject)) {
      stringOrObject = JSON.parse(stringOrObject);
    }
    if (!objects.isPlainObject(stringOrObject)) {
      return null;
    }
    const personDo = new PersonDo();
    personDo.name = this._parseString(stringOrObject, 'name');
    personDo.firstName = this._parseString(stringOrObject, 'firstName');
    personDo.dateOfBirth = this._parseDate(stringOrObject, 'dateOfBirth');
    personDo.active = this._parseBoolean(stringOrObject, 'active');
    return personDo;
  }

  protected static _parseString(o: object, property: string): string {
    return strings.asString(o[property]);
  }

  protected static _parseBoolean(o: object, property: string): boolean {
    return o ? !!o[property] : null;
  }

  protected static _parseDate(o: object, property: string): Date {
    if (!o) {
      return null;
    }
    const candidate = o[property];
    if (candidate instanceof Date) {
      return candidate;
    }
    if (objects.isString(candidate)) {
      return dates.parseJsonDate(candidate);
    }
    return null;
  }
}
